sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala

```diff
@@ -585,21 +585,26 @@ case class SortMergeJoinExec(
 
     val iterator = ctx.freshName("iterator")
     val numOutput = metricTerm(ctx, "numOutputRows")
+    val joinedRow = ctx.freshName("joined")
```

Comment:
I think we only need to do this when there is `CodegenFallback` in the condition expressions.

Reply:
The joined row should always be used for correctness. We don't know what code the expression will generate, so we should plan on always passing the correct input row. Setting left and right on a joined row is a cheap operation, so I'd rather do it correctly than rely on something brittle like `isInstanceOf[CodegenFallback]`.

Reply:
It ended up being a bit more complicated. There are two problems. The first is what this fixes, which is that the INPUT_ROW in the codegen context points to the wrong row. This is fixed and now has a test that fails if you uncomment the line that sets INPUT_ROW.

The second problem is in the check for `CodegenFallback` fails to check whether the condition supports codegen in some plans. To get the test to fail, I had to add a projection to exercise the [path where this happens](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala#L524). I'll add a second commit for this problem.

Reply:
The second problem was fixed in this commit: https://github.com/apache/spark/commit/6b6dd682e84d3b03d0b15fbd81a0d16729e521d2

I still think that the codegen problem should be fixed. Detecting `CodgenFallback` is imperfect, but will still generate code and run it. I think we should either remove codegen from `CodegenFallback` or add this fix to ensure that code works, even if we don't expect to run it.